AmplifyPlugins/DataStore/Sources/AWSDataStorePlugin/Sync/Support/APIError+Unauthorized.swift

```diff
@@ -0,0 +1,62 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Amplify
+
+extension APIError {
+
+    static let UnauthorizedMessageString: String = "Unauthorized"
+
+    /// By default, this consolidates unauthorized error scenarios coming from `APIError.httpStatusError` and
+    /// `APIError.operationError`. For `.httpStatusError`, this checks if the status code is 401 or 403. For
+    /// `.operationError`, this checks if the error description contains "Unauthorized".
+    ///
+    /// **Warning** Customized server responses that indicate unauthorized may not match the internal mapping done
+    /// in this API and return `false`. Check APIError enum directly or create your own `UnauthorizedDeterming` rule,
+    /// for example:
+    /// ```
+    /// static let customRule = UnauthorizedDetermining { error in
+    ///     // Your custom logic to determine if `error` is unauthorized.
+    ///     // return `true` or `false`.
+    /// }
+    /// ```
+    ///
+    /// - Parameter rule: Used to determine if the `APIError` is an unauthorized error.
+    /// - Returns: `true` if unauthorized error, `false` otherwise
+    public func isUnauthorized(rule: UnauthorizedDetermining = .default) -> Bool {
+        rule.isUnauthorized(self)
+    }
+
+    public struct UnauthorizedDetermining {
+        let isUnauthorized: (APIError) -> Bool
```

Comment:
Look like this came from https://github.com/aws-amplify/amplify-swift/pull/1994/files#diff-17539a8fec2423e2a770e4dff07c90e0f2e9a2678838f2c1ed139396b9ddcfad I'm trying to recall why we added this and it was because adding a new enum case to APIError was going to be a breaking change. 

I don't recall, did we decide at some point with Auth or Analytics that adding a new enum error case is no longer a major version bump? If this is the case, we can process the response within APIPlugin and directly return `APIError.unauthorized` error. This will simply things greatly

Reply:
So according to https://github.com/aws-amplify/amplify-swift#semantic-versioning-and-enumeration-cases adding APIError.unauthorized is going to be a minor version bump. I think we should decouple this additon to what you're trying to achieve here since adding a new case will require further discussion, perhaps the unauthorized case should be returned as an underlying AuthError in an APIError (one example of an alternative). 

We don't need all the public interfaces extending APIError (what happens when developers include DataStore and APIPlugin? which public `isUnauthorized` should they use?). The minimum change you will need here is just check if APIError is 401/403/"unauthorized" in operationError, which should all be internal